keras/src/distribution/tensor_parallel/autoconfig.py

```diff
@@ -0,0 +1,262 @@
+    if isinstance(layer, layers.Dense):
+        mlp_type = analyze_dense_layer(layer)
+
+        if mlp_type == "up_projection":
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=1
+            )
+            if layer.use_bias:
+                state_rules[f"{full_name}.bias"] = _split_rule(
+                    device_count, dim=0
+                )
+            output_rules[f"{full_name}"] = {0: "gather"}
+
+        elif mlp_type == "down_projection":
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=0
+            )
+            output_rules[f"{full_name}"] = {0: "allreduce"}
+
+        else:
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=1
+            )
+            if layer.use_bias:
+                state_rules[f"{full_name}.bias"] = _split_rule(
+                    device_count, dim=0
+                )
+            output_rules[f"{full_name}"] = {0: "gather -1"}
+
+    elif isinstance(layer, layers.EinsumDense):
+        if "attention_output" in full_name:
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=0
+            )
+            output_rules[f"{full_name}"] = {0: "allreduce"}
+        else:
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=1
+            )
+            if hasattr(layer, "bias") and layer.bias is not None:
+                state_rules[f"{full_name}.bias"] = _split_rule(
+                    device_count, dim=0
+                )
+            output_rules[f"{full_name}"] = {0: "gather -1"}
+
+    elif (
+        isinstance(layer, (layers.Embedding,))
+        or "Embedding" in layer.__class__.__name__
+    ):
+        if hasattr(layer, "weights"):
+            for weight in layer.weights:
+                if "embedding" in weight.name or "weight" in weight.name:
+                    key_found = False
+                    for attr_candidate in [
+                        "embeddings",
+                        "position_embeddings",
+                        "weight",
+                    ]:
+                        if getattr(layer, attr_candidate, None) is weight:
+                            state_rules[f"{full_name}.{attr_candidate}"] = (
+                                _split_rule(device_count, dim=1)
+                            )
+                            key_found = True
+                            break
+
+                    if not key_found:
+                        clean_name = weight.name.split("/")[-1].split(":")[0]
+                        state_rules[f"{full_name}.{clean_name}"] = _split_rule(
+                            device_count, dim=1
+                        )
+
+            output_rules[f"{full_name}"] = {0: "no_comm"}
+
+
+def get_default_config(module, device_ids):
+    """Generates a default tensor parallelism configuration for a Keras model.
```

Comment:
We don't use the term `module` in Keras. We have layers and models, which extend layers. I think you should just can it `model`, unless you do need to call it on layers, in which case you can call it `layer`.